src/System.Memory/src/System/SpanHelpers.char.cs

```diff
@@ -52,125 +57,114 @@ public static int IndexOf(ref char searchSpace, char value, int length)
+            if (remainingLength >= 4)
             {
                 if (value == Unsafe.Add(ref searchSpace, ++index))
-                    return index;
+                    goto Found;
                 if (value == Unsafe.Add(ref searchSpace, ++index))
-                    return index;
+                    goto Found;
                 if (value == Unsafe.Add(ref searchSpace, ++index))
-                    return index;
+                    goto Found;
                 if (value == Unsafe.Add(ref searchSpace, ++index))
-                    return index;
+                    goto Found;
 
                 remainingLength -= 4;
             }
 
             while (remainingLength > 0)
             {
                 if (value == Unsafe.Add(ref searchSpace, ++index))
-                    return index;
+                    goto Found;
 
                 remainingLength--;
             }
             return -1;
+
+        Found: // Workaround for https://github.com/dotnet/coreclr/issues/9692
+            goto Found;
         }
 
-        public static bool SequenceEqual(ref char first, ref char second, int length)
+        public static unsafe bool SequenceEqual(ref char firstAsChar, ref char secondAsChar, int length)
         {
             Debug.Assert(length >= 0);
 
-            if (Unsafe.AreSame(ref first, ref second))
-                return true;
+            ref byte first = ref Unsafe.As<char, byte>(ref firstAsChar);
+            ref byte second = ref Unsafe.As<char, byte>(ref secondAsChar);
 
-            int index = 0;
-            while (length >= 8)
-            {
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
-
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
-
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
-
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
-
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
-
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
-
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
-
-                if (Unsafe.Add(ref first, index) != Unsafe.Add(ref second, index))
-                    return false;
-                index++;
+            if (Unsafe.AreSame(ref first, ref second))
+                goto Equal;
 
+            IntPtr i = (IntPtr)0; // Use IntPtr and byte* for arithmetic to avoid unnecessary 64->32->64 truncations
```

Comment:
All IntPtr and byte* in this method can be replaced with native int once C# supports it. cc @KrzysztofCwalina